cw/cw305/capture.py

```diff
@@ -152,6 +175,102 @@ def capture_aes_random(ot, ktp):
         yield ret
 
 
+def capture_aes_random_batch(ot, ktp, capture_cfg, scope_type):
+    """A generator for capturing AES traces in batch mode.
+    Fixed key, Random texts.
+
+    Args:
+      ot: Initialized OpenTitan target.
+      ktp: Key and plaintext generator.
+      capture_cfg: Capture configuration.
+      scope_type: cw or waverunner as a scope for batch capture.
+    """
+    # Seed host's PRNG.
+    # TODO: Replace this with a dedicated PRNG to avoid other packages breaking our code.
+    random.seed(capture_cfg["batch_prng_seed"])
+    # Set the target's key
+    key = ktp.next_key()
+    tqdm.write(f'Using key: {binascii.b2a_hex(bytes(key))}')
+    cipher = AES.new(bytes(key), AES.MODE_ECB)
+    ot.target.simpleserial_write("k", key)
+    # Seed the target's PRNG
+    ot.target.simpleserial_write("s", capture_cfg["batch_prng_seed"].to_bytes(4, "little"))
+
+    # Create the ChipWhisperer project.
+    project = cw.create_project(capture_cfg["project_name"], overwrite=True)
+
+    # Capture traces.
+    rem_num_traces = capture_cfg["num_traces"]
+    num_segments_storage = 1
+    # cw and waverunner scopes are supported fot batch capture.
+    scope = SCOPE_FACTORY[scope_type](ot, capture_cfg)
+    with tqdm(total=rem_num_traces, desc="Capturing", ncols=80, unit=" traces") as pbar:
+        while rem_num_traces > 0:
+            # Determine the number of traces for this batch and arm the oscilloscope.
+            scope.num_segments = min(rem_num_traces, scope.num_segments_max)
+            scope.arm()
+
+            # Start batch encryption.
+            ot.target.simpleserial_write(
+                "b", scope.num_segments_actual.to_bytes(4, "little")
+            )
+
+            # Transfer traces
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == scope.num_segments
+            # Check that the ADC didn't saturate when recording this batch.
+            check_range(waves, ot.scope.adc.bits_per_sample)
+
+            # Generate plaintexts and ciphertexts to compare with the batch encryption results.
+            plaintexts = [ktp.next()[1] for _ in range(scope.num_segments_actual)]
+            ciphertexts = [cipher.encrypt(bytes(plaintexts[ii]))
+                           for ii in range(scope.num_segments_actual)]
```

Comment:
let's keep using `scared.aes.base.encrypt()` here.

Reply:
Is there a specific reason for that? In other functions in the same script we are using crypto.cipher and I thought it would be better if we use the same structure. In addition scared library is slower than crypto.cipher. When we try to capture 100k traces with scared we get 1697 traces/s and with crypto.cipher we get 1735 traces/s.

Reply:
I don't remember to be honest. But in general I would just change one thing at a time (merge batch script into normal script for this PR). The difference in performance you see is roughly 2%. It's not too much but maybe also not negligible. This change could be another PR. Up to now, two people have reviewed this PR and this point has been brought up by both. So please move the change into a separate PR to unblock this one.